fabtests/multinode/src/pattern.c

```diff
@@ -33,20 +33,88 @@
 #include <pattern.h>
 #include <core.h>
 
-static int pattern_next(int *cur)
+static int broadcast_gather_next(int *cur)
+{
+	int next;
+	if (pm_job.my_rank) 
+		return -FI_ENODATA;
+	next = *cur + 1;
+
+	if (next >= pm_job.num_ranks)
+		return -FI_ENODATA;
+	if (next == 0)
+		next = 1;
+
+	*cur = next;
+	
+	return 0;
+}
+
+static int broadcast_gather_current(int *cur)
+{
+	int next;
+	if (!pm_job.my_rank) 
+		return -FI_ENODATA;
+	
+	next = *cur + 1;
+
+	if (next > 0)			
+		return -FI_ENODATA;
+
+	*cur = next;
+	
+	return 0; 
+}
+
+static int ring_next(int *cur)
+{
+	if ((pm_job.my_rank == 0 && pm_job.num_ranks - 1 == *cur) ||
+		(pm_job.my_rank != 0 && pm_job.my_rank - 1   == *cur))
+		return -FI_ENODATA;
+		
+	if (pm_job.my_rank == 0)
+		*cur = pm_job.num_ranks - 1;
+	else 			
+		*cur = pm_job.my_rank - 1;
+	return 0; 
+	
+}
+
+static int ring_current(int *cur)
+{
+	if ((pm_job.my_rank + 1) % pm_job.num_ranks == *cur) 
+		return -FI_ENODATA;
+	
+	*cur = (pm_job.my_rank + 1) % pm_job.num_ranks;
+	return 0; 
+	
+}
+
+static int mesh_next(int *cur)
 {
 	int next = *cur + 1;
 
 	if (next >= pm_job.num_ranks)
-		return -ENODATA;
+		return -FI_ENODATA;
 
 	*cur = next;
 	return 0;
 }
 
+struct pattern_ops patterns[] = {
+	[0].name = "full_mesh",
+	[0].next_source = mesh_next,
+	[0].next_target = mesh_next,
+
+	[1].name = "ring",
+	[1].next_source = ring_next,
+	[1].next_target = ring_current,
+
+	[2].name = "gather",
+	[2].next_source = broadcast_gather_next,
+	[2].next_target = broadcast_gather_current,
 
-struct pattern_ops full_mesh_ops = {
-	.name = "full_mesh",
-	.next_source = pattern_next,
-	.next_target = pattern_next,
+	[3].name = "broadcast",
+	[3].next_source = broadcast_gather_current,
+	[3].next_target = broadcast_gather_next,
```

Comment:
The indices here aren't needed.  To declare this, use something like this:
```
struct pattern_ops patterns[] = {
    {
        .name = "full_mesh",
        .next_source = mesh_next,
        .next_target = mesh_next,
    },
    {
        .name = "ring",
        ...
};
```
You only want to specify the index if there's a specific location where the entry should go, in which case you would have a defined or enum value.  (E.g. [MESH_ENTRY].name = "full_mesh").

After defining the array, use:
```
#define PATTERN_COUNT ARRAY_SIZE(patterns)
```
to provide the number of patterns.  This simplifies updating the array.  You may need to define PATTERN_COUNT as a const int rather than a define, since it's used in another file.  (Or move the define to the header).